File: demos/test-ts/src/algorithms/leetcode/s0015_3sum.test.ts
function threeSum2(nums: number[]): number[][] {
  const res: number[][] = [];
  // 注意:  nums.sort() 不是按照数字大小排序的
  nums.sort((a, b) => a - b);

  for (let i = 0; i < nums.length; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) {
      continue;
    }
    let lo = i + 1;
    let hi = nums.length - 1;
    while (lo < hi) {
      const sum = nums[i] + nums[lo] + nums[hi];
      if (sum > 0) {
        hi--;
      } else if (sum < 0) {
        lo++;
      } else {
        res.push([nums[i], nums[lo], nums[hi]]);
        lo++;

        while (nums[lo] === nums[lo - 1] && lo < hi) {
          lo++;
        }
      }
    }
  }

  return res;
}

function threeSum(nums: number[]): number[][] {
  const result: number[][] = [];
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length; i++) {
    // 3数之和为 0. 第一个都大于0了。所以后面的不用比较了
    if (nums[i] > 0) {
      break;
    }
    // edge 去重
    if (i > 0 && nums[i - 1] == nums[i]) {
      continue;
    }
    let lo = i + 1;
    let hi = nums.length - 1;
    while (lo < hi) {
      const sum = nums[i] + nums[lo] + nums[hi];
      if (sum < 0) {
        lo++;
      } else if (sum > 0) {
        hi--;
      } else {
        result.push([nums[i], nums[lo], nums[hi]]);
        // edge 去重
        while (lo < hi && nums[lo + 1] == nums[lo]) {
          lo++;
        }
        // edge 去重
        while (lo < hi && nums[hi - 1] == nums[hi]) {
          hi--;
        }
        lo++;
        hi--;
      }
    }
  }

  return result;
}

test('s0015_3sum', () => {
  expect(threeSum([-1, 0])).toEqual([]);
  expect(threeSum([-1, 0, 1, 2, -1, -4])).toEqual([
    [-1, -1, 2],
    [-1, 0, 1],
  ]);
  expect(threeSum([-1, 0, 1, 2, -1, -4, -2, -3, 3, 0, 4])).toEqual([
    [-4, 0, 4],
    [-4, 1, 3],
    [-3, -1, 4],
    [-3, 0, 3],
    [-3, 1, 2],
    [-2, -1, 3],
    [-2, 0, 2],
    [-1, -1, 2],
    [-1, 0, 1],
  ]);
});

File: demos/test-go/algorithms/leetcode/s0015_3sum_test.go
package leetcode

import (
	"reflect"
	"sort"
	"testing"
)

// https://leetcode.com/problems/3sum/

func threeSum(nums []int) [][]int {
	var result [][]int

	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		if nums[i] > 0 {
			break
		}
		// edge: 跳过重复
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		lo := i + 1
		hi := len(nums) - 1
		for lo < hi {
			sum := nums[i] + nums[lo] + nums[hi]
			if sum < 0 {
				lo++
			} else if sum > 0 {
				hi--
			} else {
				result = append(result, []int{nums[i], nums[lo], nums[hi]})
				// edge: 跳过重复
				for lo < hi && nums[lo] == nums[lo+1] {
					lo++
				}
				// edge: 跳过重复
				for lo < hi && nums[hi] == nums[hi-1] {
					hi--
				}
				lo++
				hi--
			}
		}
	}

	return result
}

func Test0003(t *testing.T) {
	var tests = []struct {
		input []int
		ans   [][]int
	}{
		{[]int{-1, 0, 1, 2, -1, -4}, [][]int{{-1, -1, 2}, {-1, 0, 1}}},
	}
	for _, test := range tests {
		r := threeSum(test.input)
		if !reflect.DeepEqual(test.ans, r) {
			t.Errorf("expect %v, got %v", test.ans, r)
		}
	}
}
